SOT-131: Add Folding Scheme and R1CS definitions (#55)
sota-zk-labs/zkp-documents: terms/r1cs.md

> # R1CS

Reviewer: ```suggestion
# R1CS - Rank One Constraint System
```

## R1CS

> The R1CS structure consists of spare matrices $A, B, C$. An R1CS instance

Reviewer: ```suggestion
The R1CS structure consists of sparse matrices $A, B, C$. An R1CS instance
```

> consists of public inputs and outputs $x$ and is satisfied by a witness $W$ if:

Reviewer: ```suggestion
consists of public (inputs, outputs) $x$ and is satisfied by a witness $W$ if:
```
I think this is more clear

Author: In the definition of the paper, the author used what I wrote

$(A \cdot Z) \circ (B \cdot Z) = C \cdot Z$ where $Z = (W, x, 1)$.

For example: 

> ![r1cs](attachments/r1cs_example.png)

Reviewer 2: ```suggestion
![r1cs](./attachments/r1cs_example.png)
```

> In this picture, we consider $S$ as our assignment$Z$ above. Here, the public

Reviewer 3: math error

input and output is $out, sym_1,y, sym_2$, the witness is $x$ and $one$ is $1$.

> For further details, see this [slide](../presentations/zk_overall.pptx).

Author: R1CS is quite small to be written as an article. Maybe someone will try to convert the mentioned slide into fully documented content.

> For further details, see this [slide](../presentations/zk_overall.pptx).

Author: Vitalik didn't wrote any article about R1CS, he wrote R1CS as a part of SNARK.

## Relaxed R1CS

> The relaxed R1CS structure consists of spare matrices $A, B, C$. A relaxed

Reviewer: ```suggestion
The relaxed R1CS structure consists of sparse matrices $A, B, C$. A relaxed
```

# R1CS

## R1CS

The R1CS (Rank One Constraint System) structure consists of sparse matrices 
$A, B, C$. An R1CS instance
consists of public inputs and outputs $x$ and is satisfied by a witness $W$ if:
$(A \cdot Z) \circ (B \cdot Z) = C \cdot Z$ where $Z = (W, x, 1)$.

For example: 

> ![r1cs](./attachments/r1cs_example.png)

Reviewer 4: ```suggestion
![r1cs](attachments/r1cs_example.png)
```

> In this picture, we consider $S$ as our assignment $Z$ above. Here, the public

Reviewer 4: ```suggestion
In this picture, we consider $S$ as our assignment $Z$ above. Here, the instance - containing the public input and output - is ...
```

input and output is $out, sym_1,y, sym_2$, the witness is $x$ and $one$ is $1$.

For further details, see this [slide](../presentations/zk_overall.pptx).

## Relaxed R1CS

> The relaxed R1CS structure consists of sparse matrices $A, B, C$. A relaxed

Reviewer 4: ```suggestion
The relaxed R1CS structure consists of sparse matrices $A, B, C$. Its corresponding instance consists of
```

> R1CS instance consists of an error vector $E$, a scalar $u$ and public inputs 

Reviewer 4: ```suggestion
an error vector $E$, a scalar $u$ and public inputs 
```

and outputs $x$. An instance $(E, u, x)$ satisfied by a witness $W$ if:
$(A \cdot Z) \circ (B \cdot Z) = u \cdot (C \cdot Z) + E$, where $Z = (W, x, u)$.

Any R1CS instance can be expressed as a relaxed R1CS instance by
augmenting it with $u=1$ and $E = 0$.

## Committed Relaxed R1CS

Consider a finite field $\mathbb{F}$ and a commitment scheme $Com$ over $\mathbb{F}$.
The committed Relaxed R1CS structure consists of sparse matrices 

> $A, B, C$. A committed relaxed R1CS instance is a tuple 

Reviewer 4: ```suggestion
$A, B, C$. Its instance is a tuple 
```
